### PR TITLE
SDCORE-76 : SMF changes to support UPF allocating UE address

### PR DESCRIPTION
**Pull request by UpadhyayaAnkur**

### CODE REVIEW

File: pfcp/handler/handler.go [hunk 1]
```diff
@@ -244,13 +249,27 @@ func HandlePfcpSessionEstablishmentResponse(msg *pfcpUdp.Message) {
...
 		pfcpSessionCtx := smContext.PFCPContext[NodeIDtoIP]
 		pfcpSessionCtx.RemoteSEID = rsp.UPFSEID.Seid
 	}
 
+	//UE IP-Addr(only v4 supported)
+	if rsp.CreatedPDR != nil && rsp.CreatedPDR.UEIPAddress != nil {
+		smContext.SubPfcpLog.Infof("upf provided ue ip address [%v]", rsp.CreatedPDR.UEIPAddress.Ipv4Address)
+
+		// Release previous locally allocated UE IP-Addr
+		smContext.ReleaseUeIpAddr()
```
[the remark this answers is not in the document]
Comment (UpadhyayaAnkur) [reply pfcp/handler/handler.go:265]: We are allocating IP-Address locally at SMF beforehand, and if UPF provides any IP-Addr then we release locally allocated IP-Addr and store UPF provided IP-Addr.

File: pfcp/handler/handler.go [hunk 2]
```diff
@@ -244,13 +249,27 @@ func HandlePfcpSessionEstablishmentResponse(msg *pfcpUdp.Message) {
...
+	smContext.SMLock.Lock()
 
 	if rsp.UPFSEID != nil {
 		NodeIDtoIP := rsp.NodeID.ResolveNodeIdToIp().String()
 		pfcpSessionCtx := smContext.PFCPContext[NodeIDtoIP]
 		pfcpSessionCtx.RemoteSEID = rsp.UPFSEID.Seid
 	}
 
+	//UE IP-Addr(only v4 supported)
+	if rsp.CreatedPDR != nil && rsp.CreatedPDR.UEIPAddress != nil {
```
[the remark this answers is not in the document]
Comment (UpadhyayaAnkur) [reply pfcp/handler/handler.go:261]: So, should we check all created-PDRs to look for IP or UPF shall fill in all PDRs

File: context/datapath.go [hunk 3]
```diff
@@ -480,6 +480,15 @@ func (dpNode *DataPathNode) CreateSessRuleQer(smContext *SMContext) (*QER, error
 
 //ActivateUpLinkPdr
 func (dpNode *DataPathNode) ActivateUpLinkPdr(smContext *SMContext, defQER *QER, defPrecedence uint32) error {
+
+	ueIpAddr := pfcpType.UEIPAddress{}
+	if dpNode.UPF.IsUpfSupportUeIpAddrAlloc() {
```
Comment (thakurajayL) [inline context/datapath.go:485]: How do you do pfcp signalling when multiple UPFs ? One after the other ? Please note that logically we should get UE address from UPF and send it to iUPF... Also we should set CHV4=true only for anchor UPF

File: context/datapath.go [hunk 4]
```diff
@@ -563,6 +569,16 @@ func (dpNode *DataPathNode) ActivateUpLinkPdr(smContext *SMContext, defQER *QER,
 func (dpNode *DataPathNode) ActivateDlLinkPdr(smContext *SMContext, defQER *QER, defPrecedence uint32, dataPath *DataPath) error {
 	var iface *UPFInterfaceInfo
 	curDLTunnel := dpNode.DownLinkTunnel
+
+	//UPF provided UE ip-addr
+	ueIpAddr := pfcpType.UEIPAddress{}
+	if dpNode.UPF.IsUpfSupportUeIpAddrAlloc() {
```
Comment (thakurajayL) [inline context/datapath.go:575]: Same comment as above

File: context/datapath.go [hunk 5]
```diff
@@ -696,16 +706,21 @@ func (dataPath *DataPath) ActivateTunnelAndPDR(smContext *SMContext, precedence
 			}
 		}
 
+		ueIpAddr := pfcpType.UEIPAddress{}
+		if curDataPathNode.UPF.IsUpfSupportUeIpAddrAlloc() {
```
Comment (thakurajayL) [inline context/datapath.go:710]: only if anchor UPF

File: pfcp/handler/handler.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (thakurajayL) [reply pfcp/handler/handler.go:261]: UPF will fill in all PDRs. You can just use any one PDR

File: consumer/sm_policy.go [hunk 6]
```diff
@@ -39,7 +39,7 @@ func SendSMPolicyAssociationCreate(smContext *smf_context.SMContext) (*models.Sm
 	smPolicyData.PduSessionType = nasConvert.PDUSessionTypeToModels(smContext.SelectedPDUSessionType)
 	smPolicyData.AccessType = smContext.AnType
 	smPolicyData.RatType = smContext.RatType
-	smPolicyData.Ipv4Address = smContext.PDUAddress.To4().String()
+	smPolicyData.Ipv4Address = smContext.PDUAddress.Ip.To4().String()
```
Comment (thakurajayL) [inline consumer/sm_policy.go:42]: Can you please add intel copyright in all changed files ?

**Comment by UpadhyayaAnkur** [comment]
Changes done via another PR

**Pull request closed without merging**